### DISCUSSION TITLE
Store vectors in memory and load it on every request.

### DISCUSSION THREAD

**Discussion by gogaleko** [explosion/spaCy#6514] [Help: Coding & Implementations] [answered]
i have Georgian model, with fastText vectors. model is trained on NER. i have an API on Django.  there are several users and they have their own models and when the user send the request, i have to load his/her model and proceed data with it. loading whole model on every single request takes a long time (vectors are about 2 gb),  so i wanted to load vocab with vectors in memory once when app is started and on every request i would create new nlp (this model has "blank" vocab, its size is several kb-s, and it loads very fast) object and set vocab with vectors from memory.  see example below: 

```python

voc=Vocab().from_disk(r".\Shared\vocab")
nlp = spacy.load(modelDirectory)
nlp.vocab=voc
nlp.vocab.vectors.name = 'ge_model.vectors'

doc=nlp(u"მე მიყვარს საქართველო")
for x in doc:
    print(x.has_vector, x.vector)
``` 
this example does not works, it does not returns values.
x.has_vector is false and x.vector is empty.

But when i tried like this :

```python
nlp = spacy.load(modelDirectory)
nlp.vocab.from_disk(r".\Shared\vocab")

doc=nlp(u"მე მიყვარს საქართველო")
for x in doc:
    print(x.has_vector, x.vector)
```

x.has_vector is true and it returns vector values, but this will load this vocab every single request

also i tried load vocab and store it as bytes 

```python
voc=Vocab().from_disk(r".\Shared\vocab").to_bytes()
nlp = spacy.load(modelDirectory)

nlp.vocab.from_bytes(voc)

doc=nlp(u"მე მიყვარს საქართველო")
for x in doc:
    print(x.has_vector, x.vector)
```

this also returns values, but it still takes a lot of time.

hope i explained what i need, so is there some good way to achieve this? i searched a lot  but did not find the solution, or if someone already had the same need can you share the solution how you did it. Thanks in advance

## Your Environment

- spaCy version    2.2.4

- Platform         Windows-10-10.0.18362-SP0

- Python version   3.8.3

- Models           ge_vectors_wiki_lg, ge_wiki_lg

**Comment by adrianeboyd** [comment]
Let me make sure I understand: you have a number of NER models that were all trained with the exact same vectors and you want to be able to load the NER component in a new `nlp` object without loading the vectors from disk again?

**Comment by gogaleko** [comment]
thanks for response, yes, that's my question

**Comment by adrianeboyd** [comment, marked as answer]
If the vectors have the same name underneath (you provide it with `--vectors-name` to `spacy init-model`), you should be able to load just the NER model into a pipeline without loading the vectors again. To be sure, check that the names in `meta.json` under `["vectors"]["name"]` and in the NER model in `ner/cfg` under `"pretrained_vectors"` are the same across all your models.

Because of how the vectors are loaded in the background in v2, it's sufficient to have already loaded vectors with the same name, so you can just load the vocab without the vectors and then the NER model without the vocab into your new pipeline. Replace the empty vectors object with the one from the base model, and I think the whole `user_nlp` object should work, also for similarity and not just for the NER component:

```python
base_nlp = spacy.load("base_model_with_vectors")

user_nlp = spacy.blank("xx")
user_nlp.vocab.from_disk("user_model/vocab", exclude=["vectors"])
user_nlp.add_pipe(user_nlp.create_pipe("ner"))
user_nlp.get_pipe("ner").from_disk("user_model/ner", exclude=["vocab"])
user_nlp.vocab.vectors = base_nlp.vocab.vectors
```

There's no way to pass down the `vectors` exclude with `spacy.load()`, so this is a little involved, but I think it should work?
